[router/freebox] Plugin to monitor Freebox information

shtrom opened the pull request:
Some example graphs

![Screenshot from 2019-05-28 21-48-15](https://user-images.githubusercontent.com/160280/58475840-63bc3d80-8192-11e9-93a2-2849264f9733.png)
![Screenshot from 2019-05-28 21-48-22](https://user-images.githubusercontent.com/160280/58475841-63bc3d80-8192-11e9-83d6-ff775b1abab1.png)
![Screenshot from 2019-05-28 21-48-28](https://user-images.githubusercontent.com/160280/58475842-6454d400-8192-11e9-852e-c295125c77ef.png)

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+	curl -s "$@"
+}
+
+# XXX: testing
```
This line and the next one can probably be removed?

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+
+# XXX: testing
+# INFO="$( iconv -f latin1 -t utf8 fbx_info.txt)"
+INFO="$(fetch "${fbx_info_url}" | iconv -f latin1 -t utf8)"
```
Maybe you could provide an example of this output as a comment?

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+# INFO="$( iconv -f latin1 -t utf8 fbx_info.txt)"
+INFO="$(fetch "${fbx_info_url}" | iconv -f latin1 -t utf8)"
+
+eval $(
```
I am confused: what is the point of this `eval`?

shtrom replied:
The `sed` just create a shell snippet that defines all the variables. The `eval` brings them into the context of the current script.

sumpfralle replied:
Thank you for the clarification.
Personally I feel quite confident with writing shell scripts, but honestly I could only guess, what is happening here (due to the lack of understanding of the input text). This will be very hard for future contributors of this plugin :(
Maybe you can think of a more maintainable approach for this part?

shtrom replied:
Sadly, no. The intput text (of which I added an example above) is very non machine-parsable, so it required a lot of `sed`-trickeries to get the relevant data out of it. I can add a comment explaining the generic idea, but yeah, essentially `sed` is used to extract the data and put it in the form of a number of variable assignments, for `bash` to evaluate.

sumpfralle replied:
How about the following?
```shell
CRC_DOWN=$(echo "$INFO" | awk '/^\s+CRC\s/ { print $(NF-1); }')
CRC_UP=$(echo "$INFO" | awk '/^\s+CRC\s/ { print $NF; }')
```

Or maybe something even more readable?
```shell
get_line_column() {
    local key="$1"
    local field_index="$2"
    awk '/^\s+'"$key"'/ { print $('"$field_index"'); }'
}
...
CRC_DOWN=$(echo "$INFO" | get_line_column "CRC" "NF-1")
CRC_UP=$(echo "$INFO" | get_line_column "CRC" "NF")
```

Feel free to decide, whether such an approach could be suitable.

shtrom replied:
Ok, this mostly works, but we lose validation of the fields. My `sed` script was ensuring that the extracted data was numeric, while the `awk` way gets whatever is in the column. This means that there is some bad data in case the output file doesn't have any.

For example, with the disconnected USB data:
```USB                    Non connecté```
we now get
```usb.value Non
usb_down.value   USB                    Non connecté 
```

And I still need the `eval $(sed)` to extract the uptime.

It seem to be slightly slower, too (by 100ms or so) [this is with a local dump of the file, rather than the network]:
```
[22:40:49] ~/src/munin-contrib/plugins/router$ for i in `seq 10`; do time MUNIN_LIBDIR=.  ./freebox &> /dev/null; done                                               130 ↵ ✹ router/freebox u+1 
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.14s user 0.08s system 104% cpu 0.214 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.19s user 0.05s system 105% cpu 0.228 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.05s system 105% cpu 0.208 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.07s system 104% cpu 0.229 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.14s user 0.07s system 104% cpu 0.204 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.05s system 105% cpu 0.209 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.07s system 104% cpu 0.224 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.09s system 104% cpu 0.251 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.16s user 0.09s system 103% cpu 0.240 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.16s user 0.06s system 104% cpu 0.201 total
[22:40:54] ~/src/munin-contrib/plugins/router$ git checkout -f HEAD^                                                                                                   
HEAD is now at afb3a6e4 fixup! fixup! [router/freebox] Plugin to monitor Freebox information
[22:40:59] ~/src/munin-contrib/plugins/router$ vim freebox                                                                                                   0s  (remotes/origin/router/freebox)
[22:41:06] ~/src/munin-contrib/plugins/router$ for i in `seq 10`; do time MUNIN_LIBDIR=.  ./freebox &> /dev/null; done                                     6s  ✹ (remotes/origin/router/freebox)
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.056 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.03s system 110% cpu 0.074 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.03s user 0.03s system 112% cpu 0.052 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.06s user 0.01s system 113% cpu 0.067 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.061 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.01s system 112% cpu 0.056 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.065 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.057 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.04s user 0.02s system 112% cpu 0.059 total
```

shtrom replied:
I pushed those changes. I'll let it run for a day or so to make sure I still get all the data, and let you know.

sumpfralle replied:
Maybe the following is sufficient for numeric validation and proper handling of *unknown* values?
Bonus: only to be done once in the function :)

```shell
local result
result=$(awk '/^\s+'"$key"'/ { print $('"$field_index"'); }')
if echo "$result" | grep -q '^[0-9]\+$'; then
    echo "$result"
else
    echo "U"
fi
```

Regarding the uptime:
```shell
UPTIME_DAYS=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-5")
UPTIME_HOURS=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-3")
UPTIME_MINUTES=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-1")
```

Maybe I am a bit obsessed about code readability :)

shtrom replied:
Hum, I suppose that could work for the uptime. As for the validation, would there be a more direct way to match the regexp straight within `awk`? The `grep` should work, but it feels like a lot of overhead considering we already parsed the data out.

shtrom replied:
Ok, I found a way. How about now?

sumpfralle replied:
Yes, this should work easily by prepending the `print` call with something like `if ($('"$field_index"') =~ /^[0-9]+$/)`.

sumpfralle replied:
This looks good to me!
Just one thing: maybe you need to add `^` in front of `/[.0-9]+/`? (otherwise `foo4` would match?)

Thank you for bearing with my wishes! I am happy to merge it: just say so.

shtrom replied:
Ah, good catch on the `^...$` in the RE.

It's all fixed now, and squashed, so should be ready to merge :+1: .

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+
+DHCP_CLIENTS=$(sed -nE '/Attributions dhcp/,/Redirections de ports/{s/^\s*([0-9A-F:]{17}).*$/\1/p}' <<< "${INFO}" | wc -l)
+REDIRECT_TCP=$(sed -nE '/Redirections de ports/,/Interfaces réseau/{s/^\s*(TCP).*$/\1/p}' <<< "${INFO}" | wc -l)
+REDIRECT_UDP=$(sed -nE '/Redirections de ports/,/Interfaces réseau/{s/^\s*(UDP).*$/\1/p}' <<< "${INFO}" | wc -l)
```
The `sed` command `for `REDIRECT_UDP` looks like it would count the sum of the TCP and UDP redirects (since the start line pattern is equal).

shtrom replied:
Nope, both are in the same section (`Redirection de ports`), but the `s///p` outputs only lines with matching replacements (so, either `TCP` or `UDP`).

sumpfralle replied:
OK - thank you.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+REDIRECT_TCP=$(sed -nE '/Redirections de ports/,/Interfaces réseau/{s/^\s*(TCP).*$/\1/p}' <<< "${INFO}" | wc -l)
+REDIRECT_UDP=$(sed -nE '/Redirections de ports/,/Interfaces réseau/{s/^\s*(UDP).*$/\1/p}' <<< "${INFO}" | wc -l)
+
+FREEBOX_UPTIME=$(echo "${UPTIME_DAYS}*86400+${UPTIME_HOURS}*3600+${UPTIME_MINUTES}*60" | bc -ql)
```
Maybe `FREEBOX_UPTIME=$((UPTIME_DAYS * 86400 + UPTIME_HOURS * 3600 + UPTIME_MINUTES * 60))` would be more readable (and not require `bc`)?

shtrom replied:
It would be, wouldn't it. I've always assumed that it was a `bash`ishm, but it actually seems to work with `sh`... TIL.

shtrom replied:
And I'm using `bash` anyway :man_facepalming:

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+
+=head1 NAME
+
+freebox - Plugin to monitor stats of a Freebox
```
Maybe you could provide a bit of details, what this `Freebox` is?

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+
+freebox - Plugin to monitor stats of a Freebox
+
+The ideal usage is also used as an updated warning limit.
```
Could you rephrase this, please?
(I do not understand it spontaneously)

shtrom replied:
Ah, this was from a different plugin I took as a base to work on this one. It's not relevant here.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+LAST_CONNECT=$(sed -nE '/Journal de connexion adsl/,+5{s#^\s*([0-9]{2})/([0-9]{2})/([0-9]{4})[^0-9]*([0-9]+):([0-9]+):([0-9]+)\s+Connexion.*#\3-\2-\1T\4:\5:\6#p}' <<< "${INFO}")
+CONNECTION_UPTIME=na
+if [ -n "${LAST_CONNECT}" ]; then
+	FORMULA="$(date +%s)-$(date -d "${LAST_CONNECT}" +%s)"
```
I could imagine, that the following is as readable as your suggestion (without requiring `bc`):
```
FORMULA=$(( $(date +%s) - $(date -d "$LAST_CONNECT" + %s) ))

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+			echo 'graph_vlabel bits per second in (+) / out (-)'
+			echo 'graph_order main_atm_down main_atm main_wan_down main_wan'
+
+			# XXX: parameters duplicated from traffic
```
Why are these duplicated?

shtrom replied:
It's a multigraph, so I have those both in the traffic subgraph, and the main graph.

I'm curious to know if there is any other way to do this.

sumpfralle replied:
Ah - ok. I did not get, that the duplication is in the main graph. This is indeed fine.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+			echo 'redirect_tcp.label TCP redirections'
+			echo 'redirect_udp.label UDP redirections'
+			;;
+		*)
```
Is this supposed to be the graph `''` (empty string)? How about using `''` (two single quotes) instead?

shtrom replied:
It's a catch-all, I prefer to have that, to always output something, rather than be more limitative. Granted, there is little chance that anything else that `''` will ever get there, but I find it safer.

sumpfralle replied:
I am more of a python person - thus I like it, when the program informs me when an older assumption is not true anymore. Anyway, it is just a matter of style ...

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+    set -x
+fi
+
+for CMD in curl iconv bc; do
```
Please move this check to `autoconf`. Now it would cause the `autoconf` call to fail. This should not happen.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+			if curl -qso /dev/null $fbx_info_url; then
+				echo 'yes'
+			else
+				echo 'no'
```
Please provide an explanation for `no`, e.g.: `no (failed to retrieve '$fbx_info_url')`

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+main() {
+	case $1 in
+		autoconf)
+			if curl -qso /dev/null $fbx_info_url; then
```
Please quote the URL.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,316 @@
+
+}
+
+main $1
```
Please quote `$1`. Maybe you even want to use `main "${1:-}"`.
In this case you should almost be able to enable "error on unset" (via `set -u`).

shtrom commented:
@sumpfralle good for another look. I left the fixup commit there with all the changes for the moment, to ease incremental reviews; I'll squash at the end.

sumpfralle commented on plugins/router/freebox:
```diff
@@ -0,0 +1,456 @@
+	echo "multigraph freebox${graph}"
+	case "$graph" in
+		.adsl)
+			echo "noise.value ${NOISE_UP:-nan}"
```
Instead of `nan` you should use `U` (unknown value).

shtrom commented:
@sumpfralle done: I added a comment with an example about the `eval $(sed ...)`, and replaced all `nan`s with `U` (thanks for that, I have been looking for the correct placeholder for a while, but couldn't find it anywhere.

If you're happy to merge, feel free to squash the fixups first.

sumpfralle commented:
Thank you for your patience!
